### ISSUE TITLE
libdislocator breaking change: getrandom not available on older Linux kernels

### ISSUE THREAD

**Issue by richinseattle** [vanhauser-thc/AFLplusplus#123]
I have commented on the commit, but the latest update to libdislocator creates a dependency on the getrandom system call which does not exist in older kernels, such as 3.16 which ships with Debian 9 (which is EOL next June). 

https://github.com/vanhauser-thc/AFLplusplus/commit/93a9e2daf34dd37fdd44f4239b14fe29f8604da2#r36207264

**Comment by andreafioraldi** [comment]
Yes, I see, I'll fix ASAP.
@vanhauser-thc we should test PRs against these issues with legacy systems before accepting them cause many people fuzz on those systems (especially embedded).

**Comment by andreafioraldi** [comment]
Oh https://github.com/vanhauser-thc/AFLplusplus/pull/124 should fix the issue, @richinseattle can quicly you test it for us on your Debian?

**Comment by vanhauser-thc** [comment]
I already accepted the PR as it improves the situation.
not sure its complete though. someone can have an old kernel but a new glibc, would that still break?
Not sure 

@devnexen  ?

**Comment by andreafioraldi** [comment]
IMO this fix is better:
```c
#ifdef __linux__
  #define _GNU_SOURCE
  #include <unistd.h>
  #include <sys/syscall.h>
  #ifdef __NR_getrandom
    #define arc4random_buf(p, l) do { ssize_t rd = syscall(__NR_getrandom, p, l, 0); if (rd != l) DEBUGF("getrandom failed"); } while(0)
  #else
    #include <time.h>
    #define arc4random_buf(p, l) do { srand(time(NULL)); u32 i; u8 *ptr = (u8 *)p; for(i = 0; i < l; i++) ptr[i] = rand() % INT_MAX; } while(0)
  #endif
#endif
```

**Comment by devnexen** [comment]
slighty better solution : 
`#if defined(SYS_getrandom) && defined(__NR_getrandom)` 
to be sure the syscall id has been defined.

**Comment by richinseattle** [comment]
The current patch is incorrect.. it uses || instead of && on the minor version, but I think Andrea's approach is better since the macro is referencing the syscall number directly. Changing the logical OR to AND fixes it, as does replacing with Andrea's patch. 

Extra environment info I should have included the first time (note I symlinked AFLplusplus to afl hence the paths):

vulndev@vulndev:/vulndev/afl/libdislocator$ uname -a 
Linux vulndev 3.16.0-6-amd64 #1 SMP Debian 3.16.56-1+deb8u1 (2018-05-08) x86_64 GNU/Linux
vulndev@vulndev:/vulndev/afl/libdislocator$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
vulndev@vulndev:/vulndev/afl/libdislocator$ ldd --version 
ldd (Debian GLIBC 2.19-18+deb8u10) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

**Comment by vanhauser-thc** [comment]
I don't have an old linux in a docker at the moment plus I am traveling so I cant really help here at the moment :( so I just accept PRs ...

I accepted the revised patch from @devnexen 

@richinseattle does it work now?

@andreafioraldi does it look ok? if not can you push a change?

I put it on my plate to make a travis file that runs on bionic plus an older platform (@richinseattle : trusty? xenial? what is your opinion?)

@andreafioraldi we can also build on different CPU: amd64, ppc64le, s390x, arm64 - is that something you would like to have in there?

@hexcoder- travis fails as long as gcc_plugin instrumentation is not improved (and explicitly inserted a testcase to test for that). are you still on vacation or back? this needs to be addressed either by being ?fixed/workaround/..? or that we move gcc_plugin back to a branch. in its current state its worse than afl-gcc, even with whitelists enabled (not your fault, gcc plugins are crap IMHO from what I played around with it).

**Comment by andreafioraldi** [comment]
Differy CPUs would be useful especially for QEMU but it may cost too much. radare2 was banned more than one time IIRC from travis due to the too long build time. Maybe just the two most used, x86 and ARM, is fine.
I'll push a fix to the fix of the fix, sys/syscall.h has to be included before the ifdef not after.

**Comment by hexcoder-** [comment]
It might be necessary to move the gcc plugin pass more to a later place in the chain. More investigation seems necessary.

**Comment by vanhauser-thc** [comment]
@hexcoder- I also notice large differences between gcc versions. gcc-7 seems to be OK for example, gcc-9 produces too many instrumentations.

@all: I added other travis targets. just cost 3 hours of my life ... *sigh*
we now have: Ubuntu 18.04/bionic, 16.04/xenial and 14.04 trusty, plus we have an Ubuntu 16.04/xenial on ARM64. They sometimes fail on radamsa, that it does not find solutions for afl-fuzz. increased the time for that testcase and hope it works now more reliable ...
I also tried to get OS/X working with travis, but afl only works on OS/X if a custom llvm and gcc is installed (with brew or direct binary package), and brew takes too long for travis. so no OS/X target for us.
